node/cmd/ccq/p2p.go

```diff
@@ -0,0 +1,237 @@
+				// This intentionally makes us incompatible with the global IPFS DHT
+				dht.ProtocolPrefix(protocol.ID("/"+networkID)),
+				dht.BootstrapPeers(bootstrappers...),
+			)
+			return idht, err
+		}),
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	topicName := fmt.Sprintf("%s/%s", networkID, "broadcast")
+
+	logger.Info("Subscribing pubsub topic", zap.String("topic", topicName))
+	ps, err := pubsub.NewGossipSub(ctx, h)
+	if err != nil {
+		return nil, err
+	}
+
+	topic, err := ps.Join(topicName)
+	if err != nil {
+		logger.Error("failed to join topic", zap.Error(err))
+		return nil, err
+	}
+
+	sub, err := topic.Subscribe()
+	if err != nil {
+		logger.Error("failed to subscribe topic", zap.Error(err))
+		return nil, err
+	}
+
+	logger.Info("Node has been started", zap.String("peer_id", h.ID().String()),
+		zap.String("addrs", fmt.Sprintf("%v", h.Addrs())))
+
+	// Wait for peers
+	for len(topic.ListPeers()) < 1 {
+		time.Sleep(time.Millisecond * 100)
+	}
+
+	// Fetch the initial current guardian set
+	guardianSet, err := FetchCurrentGuardianSet(ethRpcUrl, ethCoreAddr)
+	if err != nil {
+		logger.Fatal("Failed to fetch current guardian set", zap.Error(err))
+	}
+	quorum := vaa.CalculateQuorum(len(guardianSet.Keys))
+
+	// Listen to the p2p network for query responses
+	go func() {
+		// Maps the request signature to a map of response digests which maps to a list of guardian signatures.
+		// A request could have responses with different digests, because the guardians could have
+		// different results returned for the query in the event of a rollback.
+		responses := make(map[string]map[ethCommon.Hash][]GuardianSignature)
+		for {
+			envelope, err := sub.Next(ctx)
+			if err != nil {
+				logger.Error("Failed to read next pubsub message", zap.Error(err))
+				break
+			}
+			var msg gossipv1.GossipMessage
+			err = proto.Unmarshal(envelope.Data, &msg)
+			if err != nil {
+				logger.Error("received invalid message", zap.Binary("data", envelope.Data),
+					zap.String("from", envelope.GetFrom().String()))
+				continue
+			}
+			switch m := msg.Message.(type) {
+			case *gossipv1.GossipMessage_SignedQueryResponse:
+				logger.Debug("query response received", zap.Any("response", m.SignedQueryResponse))
+				var queryResponse query.QueryResponsePublication
+				err := queryResponse.Unmarshal(m.SignedQueryResponse.QueryResponse)
+				if err != nil {
+					logger.Error("failed to unmarshal response", zap.Error(err))
+					continue
+				}
+				requestSignature := hex.EncodeToString(queryResponse.Request.Signature)
+				// Check that we're handling the request for this response
+				pendingResponse := pendingResponses.Get(requestSignature)
+				if pendingResponse == nil {
+					logger.Debug("skipping query response for unknown request", zap.String("signature", requestSignature))
+					continue
+				}
+				// Make sure that the request bytes match
+				if !bytes.Equal(queryResponse.Request.QueryRequest, pendingResponse.req.QueryRequest) ||
```

Comment:
Is it possible that some request gets stuck in here from a previous error leg and all future responses get passed over because the requests don't match?

Reply:
The request signature is removed from the `pendingResponses` container after the http request is handled (either successfully or an error) so future requests with the same signature shouldn't get stuck